### DISCUSSION TITLE
Add Built-in Soft Delete Functionality

### DISCUSSION THREAD

**Discussion by HoangMinhBK** [sqlalchemy/sqlalchemy#11468] [Ideas]
### Describe the use case

I propose adding built-in support for soft deletes in SQLAlchemy. This feature would allow users to mark records as deleted without physically removing them from the database.

For example, in a customer management system, you might want to "delete" a customer who has requested account closure. However, for audit purposes and potential future recovery, you don't want to remove their data completely from the database. Instead, you mark the customer as deleted, which prevents the data from appearing in regular queries but keeps it available for administrators or recovery operations.

### Databases / Backends / Drivers targeted

- **Databases**: PostgreSQL, MySQL, SQLite, and any other database that supports Boolean fields.
- **Drivers**: All drivers supported by SQLAlchemy (e.g., psycopg2 for PostgreSQL, pymysql for MySQL).

### Example Use

**SQL Example**
Here's what the SQL might look like for a soft delete operation:
```
-- Mark a record as deleted
UPDATE some_model SET deleted = TRUE WHERE id = 1;

-- Query non-deleted records
SELECT * FROM some_model WHERE deleted = FALSE;
```
**DBAPI Example**
Using SQLAlchemy to perform a soft delete and query non-deleted records:
```
from sqlalchemy.ext.declarative import declared_attr
from sqlalchemy import Boolean, Column, Integer, String, create_engine, inspect
from sqlalchemy.orm import sessionmaker, scoped_session, declarative_base

Base = declarative_base()

class SoftDeleteMixin:
    @declared_attr
    def deleted_column(cls):
        return 'deleted'

    @declared_attr
    def deleted(cls):
        return Column(Boolean, default=False)

    def soft_delete(self):
        if not hasattr(self, self.deleted_column):
            raise AttributeError(f"This record does not have '{self.deleted_column}' column to support soft delete.")

        setattr(self, self.deleted_column, True)

# Example Model using the SoftDeleteMixin with an alias
class SomeModel(Base, SoftDeleteMixin):
    __tablename__ = 'some_model'
    id = Column(Integer, primary_key=True)
    name = Column(String(50), nullable=False)
    deleted_column = 'is_deleted'  # Alias for the deleted field

# Create a new record
new_record = SomeModel(name="Example")
session.add(new_record)
session.commit()

# Soft delete the record
record = session.query(SomeModel).first()
try:
    record.soft_delete()
    session.commit()
except AttributeError as e:
    print(e)

# Recover soft deleled record
    def recover(self):
        if not hasattr(self, self.deleted_column):
            raise AttributeError(f"This record does not have '{self.deleted_column}' column for soft delete.")
        setattr(self, self.deleted_column, False)

# Query non-deleted records
non_deleted_records = session.query(SomeModel).filter(getattr(SomeModel, SomeModel.deleted_column) == False).all()
print(non_deleted_records)
```

### Additional context

The proposed implementation includes:

1. **SoftDeleteMixin**: A mixin that adds a `deleted_column` can be aliased into `deleted`, `is_deleted` (user choice) and a `soft_delete` method. This method raises Attributes error if the record doesn't a `deleted_column` used for soft delete.
3. **Recovery Support**: Methods to recover soft-deleted records.

The soft delete functionality will help standardize a common pattern.

**Comment by hasansezertasan** [comment]
Related or duplicate of: #3596, #4004, and #7973

I also would like to have this version along with an automatic history generator version like sqlalchemy-continuum.

**Comment by zzzeek** [comment]
hi -

You have seen issues #3596, #4004, #7973.   My answers on those issues for the most part still stand, this is something we can document further, provide more examples, but is not appropriate as a built in "feature" of SQLAlchemy that we'd have to support in perpetuity, including supporting any and all edge cases which can make what would normally be simple end-user recipes into vastly overcomplex and difficult-to-use builtins.

**Reply by awoimbee** [reply]
> My answers on those issues for the most part still stand

Meaning you still think: https://github.com/sqlalchemy/sqlalchemy/issues/7973#issuecomment-1112561295
> the soft delete pattern in SQLAlchemy 1.4 is now provided by the do_orm_execute() event in conjunction with the with_loader_criteria() ORM option.

Which AFAIK doesn't work if you want to load a non hidden orm object which has a relationship that should load hidden objects. 

---

I have this monstrosity of a query mixin that I don't see how to migrate to sqlalchemy v2:
```python
class WithHiddenQueryMixin[T: DeclarativeBase]:
    """Extend queries to automatically filter out hidden elements from them."""

    _with_hidden = False

    def __new__(cls, *args, **kwargs):
        obj = super(WithHiddenQueryMixin, cls).__new__(cls)
        obj._with_hidden = False
        if len(args) > 0:
            super(WithHiddenQueryMixin, obj).__init__(*args, **kwargs)
            return obj.filter_by(hidden=False)
        return obj

    def with_hidden(self, strict=True) -> Self:
        """Override behaviour on query creation by using this method."""
        # Remove the filter_by(hidden=false) we setup in __new__
        if self._where_criteria and self._where_criteria[0].left.name == "hidden":
            self._where_criteria = self._where_criteria[1:]
        elif strict:
            raise RuntimeError(
                "with_hidden called on a query with no hidden=False filter"
            )
        self._with_hidden = True
        return self

    def _get(self, *args, **kwargs) -> T | None:
        """Original query.get from base class."""
        return super().get(*args, **kwargs)

    def get(self, *args, **kwargs) -> T | None:
        """Workaround for get not handling filters being pre-loaded."""
        with_hidden = self._with_hidden
        obj = self.with_hidden(strict=False)._get(*args, **kwargs)
        if obj is not None and (with_hidden or not obj.hidden):
            return obj
        return None
```

**Reply by CaselIT** [reply]
it's not really clear what the aim of that code is, can you open a new discussion explaining what you are trying to do with an example in addition of your code?
